src/persist-client/src/internal/state.rs

```diff
@@ -1318,12 +1337,64 @@ where
             K::decode_schema(&x.key) == *key_schema && V::decode_schema(&x.val) == *val_schema
         });
         match existing_id {
-            Some((schema_id, _)) => {
-                // TODO: Validate that the decoded schemas still produce records
-                // of the recorded DataType, to detect shenanigans. Probably
-                // best to wait until we've turned on Schema2 in prod and thus
-                // committed to the current mappings.
-                Break(NoOpStateTransition(Some(*schema_id)))
+            Some((schema_id, encoded_schemas)) => {
+                let schema_id = *schema_id;
+                let new_k_datatype = mz_persist_types::columnar::data_type::<K>(key_schema)
+                    .expect("valid key schema");
+                let new_v_datatype = mz_persist_types::columnar::data_type::<V>(val_schema)
+                    .expect("valid val schema");
+
+                let new_k_encoded_datatype = encoded_data_type(&new_k_datatype);
+                let new_v_encoded_datatype = encoded_data_type(&new_v_datatype);
+
+                // Check if the generated Arrow DataTypes have changed.
+                if encoded_schemas.key_data_type != new_k_encoded_datatype
+                    || encoded_schemas.val_data_type != new_v_encoded_datatype
+                {
+                    let old_k_datatype =
+                        decode_data_type(Bytes::clone(&encoded_schemas.key_data_type))
+                            .expect("failed to roundtrip Arrow DataType");
+                    let old_v_datatype =
+                        decode_data_type(Bytes::clone(&encoded_schemas.val_data_type))
+                            .expect("failed to roundtrip Arrow DataType");
+
+                    let k_atleast_as_nullable =
+                        crate::schema::is_atleast_as_nullable(&old_k_datatype, &new_k_datatype);
+                    let v_atleast_as_nullable =
+                        crate::schema::is_atleast_as_nullable(&old_v_datatype, &new_v_datatype);
+
+                    // If the Arrow DataType for `k` or `v` has changed, but it's only become more
+                    // nullable, then we allow in-place re-writing of the schema.
```

Comment:
Maybe this is friday brain, but: since we're updating existing schema ids here with a backwards compatible change, what would go wrong if we didn't do the schema-id-deprecation thing? Or nothing breaks, and we're doing it to avoid ambiguity?

Reply:
Not Friday brain at all, there are a couple of trade-offs here. We could skip deprecating the `schema_id` field and just write a new `SchemaId(1)` for all collections, but:
* We would have have the same `RelationDesc` mapping to two different `arrow::DataType`s, which I don't think is too bad, but would probably require us to also implement something like https://github.com/MaterializeInc/materialize/pull/30627 in the future.
* Right now the Persist `SchemaId` maps 1:1 with the Adapter's `RelationVersion` for Tables. If we evolve the schemas for all collections to be more nullable we'd need some additional mapping in the Adapter layer to support `ALTER TABLE`

These are all things that we can work around, but also IMO the concept of "starting" over is the cleanest choice.